### PR TITLE
Change the `?` lowering to selectively omit the `from`

### PR DESCRIPTION
**Pull request by llogiq**
This omits the `From::from(_)` call around the error expression if and only if there is no catch block and the surrounding function or method returns an `impl Trait` type. This should help type inference to figure out the correct type.

Since we change the meaning of the code, this probably requires a crater run before merging.

### CODE REVIEW

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -5214,3 +5235,12 @@ fn body_ids(bodies: &BTreeMap<hir::BodyId, hir::Body>) -> Vec<hir::BodyId> {
     body_ids.sort_by_key(|b| bodies[b].value.span);
     body_ids
 }
+
+fn returns_impl_trait(decl: &FnDecl) -> bool {
+    if let FunctionRetTy::Ty(ref ty) = decl.output {
+        if let TyKind::ImplTrait(..) = ty.node {
```
Comment (oli-obk) [inline src/librustc/hir/lowering.rs]: I think we should also support existential types here. Existential types should be detectable by checking which definition a `TyKind::Path` refers to. Might require a lowered function signature

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -95,6 +95,9 @@ pub struct LoweringContext<'a> {
     modules: BTreeMap<NodeId, hir::ModuleItems>,
 
     is_generator: bool,
+    /// This is true iff we are in a function that returns an impl Trait
```
Comment (ljedrz) [inline src/librustc/hir/lowering.rs]: nit: iff

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -4555,11 +4572,15 @@ impl<'a> LoweringContext<'a> {
                 let err_arm = {
                     let err_ident = self.str_to_ident("err");
                     let (err_local, err_local_nid) = self.pat_ident(e.span, err_ident);
-                    let from_expr = {
+                    let err_expr = self.expr_ident(e.span, err_ident, err_local_nid);
+                    // we omit the `From::from` if there is no catch scope and we are
+                    // in an `impl trait` returning function
```
Comment (ljedrz) [inline src/librustc/hir/lowering.rs]: nit: trait could be capitalized

File: src/librustc/hir/lowering.rs (hunk 2 above)
Comment (llogiq) [reply src/librustc/hir/lowering.rs]: Ok, I should unabbreviate if-and-only-if.

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -4555,11 +4572,15 @@ impl<'a> LoweringContext<'a> {
...
+                    let from_expr = if if self.catch_scopes.is_empty() && self.returns_impl {
```
Comment (scottmcm) [inline src/librustc/hir/lowering.rs]: ` if if `?

**Comment by scottmcm** [comment]
Please include tests showing what it can now do.

I suspect that just removing the `from` call is insufficient, since the desugar will still call `Try::from_error`, which depends on context to determine the return type.

Considering posting ideas you have in https://github.com/rust-lang/rust/issues/42327 -- `Try` may end up changing, and this kind of desire would be a good thing to capture there.

File: src/librustc/hir/lowering.rs [hunk 5]
```diff
@@ -4555,11 +4572,15 @@ impl<'a> LoweringContext<'a> {
...
                     let err_ident = self.str_to_ident("err");
                     let (err_local, err_local_nid) = self.pat_ident(e.span, err_ident);
-                    let from_expr = {
+                    let err_expr = self.expr_ident(e.span, err_ident, err_local_nid);
+                    // we omit the `From::from` if there is no catch scope and we are
+                    // in an `impl Trait` returning function
+                    let from_expr = if self.catch_scopes.is_empty() && self.returns_impl {
+                        err_expr
+                    } else {
                         let path = &["convert", "From", "from"];
```
Comment (Centril) [inline src/librustc/hir/lowering.rs]: Why is `From` not a lang item?...

**Comment by Centril** [comment]
This changes behavior on stable and so most definitely needs approval from the language team. I also think that while the definition of `Try` is in limbo, we should not do such a change (if at all, because it makes the semantics of `?` more contextual and harder to reason about) since it carries hard to tell risks for the future and I really cannot judge the long term impact of this change.

**Comment by oli-obk** [comment]
We could solely support it for existential types only for now and block their stabilization of it on it working nicely there (or remove it again if we realize it's unfeasible)

**Comment by Centril** [comment]
@oli-obk Could it be gated separately perhaps? A change this small even for `existential type` is small enough that it could accidentally slip into stabilization (in the sense that I know I could make this mistake myself on a bad day) when we do it for `existential type` (also, the tracking issue is https://github.com/rust-lang/rust/issues/34511 which is a complete mess so any note in the issue description would also be easily missed).

**Comment by CryZe** [comment]
Potential problems with stability aside, I'm confused why the implementation was only attempted for impl Trait return types in the first place, this problem affects existential types and normal generics too. So this implementation only working for very specific patterns seems very inconsistent.

**Comment by llogiq** [comment]
I started with the minimal implementation that only looks for impl Trait because that's what we use in doctests, where I found the resulting type inference papercut.

So, I'll gladly extend this to all existential types if it makes it easier to reason about.

**Comment by llogiq** [comment]
@oli-obk @CryZe how do I check if a feature gate is active? And how do I check for an existential return type? Is it sufficient to check for `TyKind::TraitObject`, `TyKind::Path` and `TyKind::Def` with non-empty `GenericArg`s?

**Comment by oli-obk** [comment]
> how do I check if a feature gate is active?

`tcx.features().feature_gate_name`

> And how do I check for an existential return type?

Probably `TyKind::Path` and check if that path is a local one and if the item it points to is an existential type declaration.

**Comment by llogiq** [comment]
Ok, so I'll put this behind a feature gate and add a test for the impl trait variant. @CryZe @oli-obk does someone have examples how to create a generic or existential `Try` type so I can add tests for those variants, too?

**Comment by llogiq** [comment]
Ok, so I'll put this behind a feature gate and add a test for the impl trait variant. @CryZe @oli-obk does someone have examples how to create a generic or existential `Try` type so I can add tests for those variants, too?

**Comment by oli-obk** [comment]
If you just have a session available, you can use https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.features_untracked

> does someone have examples how to create a generic or existential Try type so I can add tests for those variants, too?

basically use the same test you did for `impl Trait`, but replace the return type with `Foo` and create an `existential type Foo: Trait;` item somewhere in the same module.

**Comment by cramertj** [comment]
I'm not sure I understand this PR. My expectation would be that this would compile:

```rust
fn try_thing() -> Result<(), std::io::Error> { ... }
fn try_other_thing() -> Result<(), failure::Error> { ... }

existential type X:;
fn foo() -> X {
    try_thing()?;
    try_other_thing()?;
    Ok::<(), failure::Error>(())
}
```

**Comment by llogiq** [comment]
@cramertj thanks for coming up with a counter example. Seems like putting the cleverness into typeck somehow might be the way to go here. Closing this.

**Comment by glaebhoerl** [comment]
(Related-seemingly: back on the tracking issue I [suggested](https://github.com/rust-lang/rust/issues/42327#issuecomment-382407423) (with an excess of throat-clearing) that when `From::from` in the `?` sugaring would lead to an ambiguity, we default to selecting `impl<T> From<T> for T`, that is leaving the type unchanged.)

**Comment by cramertj** [comment]
@glaebhoerl I'd love that, but I also have no idea how to implement it (in general, "if this doesn't compile then try X"-style checks are hard to make work). Today we insert the `into()` conversion during lowering to HIR, which is long before we'd have enough information to determine if an inference failure would result.

**Comment by llogiq** [comment]
Perhaps we could somehow annotate the `From` and `Into` traits with some hint for infer on the expected type?

**Comment by glaebhoerl** [comment]
I was initially thinking it'd be analogous to defaulting for types (e.g. of integer literals to `i32`), just for `impl`s, but maybe that wouldn't work if impl resolution doesn't work by representing the to-be-resolved `impl`s as variables. (I have no idea whether defaulting is something that could fit well into however impl resolution *does* work.)

On the other hand, another way of achieving the same result which *does* work using defaulting of type/inference variables (which, again, we already have), is that if a variable is ambiguous but is bounded by `From`/`Into` where the other type is known, then default the variable to the same type. This will then result in the `From<T> for T` impl being selected.

(I am not familiar with the details of the typechecking algorithm to evaluate how viable that is, but it seems at least *plausible*.)

**Pull request closed without merging**